File: eel-core/src/main/scala/io/eels/component/avro/AvroSource.scala
package io.eels.component.avro

import java.io.File
import java.util.concurrent.atomic.AtomicBoolean

import com.sksamuel.exts.Logging
import com.sksamuel.exts.io.Using
import io.eels._
import io.eels.datastream.{DataStream, Publisher, Subscriber, Subscription}
import io.eels.schema.StructType
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileSystem, Path}

case class AvroSource(path: Path)
                     (implicit conf: Configuration, fs: FileSystem) extends Source with Using {

  override lazy val schema: StructType = {
    using(AvroReaderFns.createAvroReader(path)) { reader =>
      val record = reader.next()
      AvroSchemaFns.fromAvroSchema(record.getSchema)
    }
  }

  override def parts(): Seq[Publisher[Seq[Row]]] = Seq(AvroSourcePublisher(path))
}

case class AvroSourcePublisher(path: Path)
                              (implicit conf: Configuration, fs: FileSystem)
  extends Publisher[Seq[Row]] with Logging with Using {
  override def subscribe(subscriber: Subscriber[Seq[Row]]): Unit = {
    val deserializer = new AvroDeserializer()
    try {
      using(AvroReaderFns.createAvroReader(path)) { reader =>
        val running = new AtomicBoolean(true)
        subscriber.subscribed(Subscription.fromRunning(running))
        AvroRecordIterator(reader)
          .takeWhile(_ => running.get)
          .map(deserializer.toRow)
          .grouped(DataStream.DefaultBatchSize)
          .foreach(subscriber.next)
        subscriber.completed()
      }
    } catch {
      case t: Throwable => subscriber.error(t)
    }
  }
}

object AvroSource {
  def apply(file: File)(implicit conf: Configuration, fs: FileSystem): AvroSource = AvroSource(new Path(file.getAbsoluteFile.toString))
  def apply(path: java.nio.file.Path)(implicit conf: Configuration, fs: FileSystem): AvroSource = apply(path.toFile)
}

File: eel-core/src/main/kotlin/io/eels/component/avro/AvroSource.kt
package io.eels.component.avro

import io.eels.Row
import io.eels.Source
import io.eels.component.Part
import io.eels.component.Using
import io.eels.schema.Schema
import rx.Observable
import java.nio.file.Path

class AvroSource(val path: Path) : Source, Using {

  override fun schema(): Schema {
    return using(createAvroReader(path), { reader ->
      val record = reader.next()
      fromAvroSchema(record.schema)
    })
  }

  override fun parts(): List<Part> = listOf(AvroSourcePart(path, schema()))
}

class AvroSourcePart(val path: Path, val schema: Schema) : Part {

  override fun data(): Observable<Row> = Observable.create<Row> { subscriber ->

    val reader = createAvroReader(path)
    subscriber.onStart()

    while (reader.hasNext() && !subscriber.isUnsubscribed) {
      val record = reader.next()
      val row = AvroRecordDeserializer().toRow(record)
      subscriber.onNext(row)
    }

    if (!subscriber.isUnsubscribed)
      subscriber.onCompleted()

    reader.close()
  }
}